---
title: "ATAC-seq Report"
output: pdf_document
params:
  gene: NA
  ymax: NA
  xrange: NA
  gvizPlot: NA
  tf_legend: NA
  rnaPlot: NA
  transcript_start: NA
  transcript_end: NA
  chr: NA
---

```{r include=FALSE}
library(Gviz)
library(ggplot2)
library(ggpubr)
```

## Summary for `r params$gene`.
Include information such as correlation cuttoff, selected transcript, etc.

\newpage

```{r echo=FALSE, fig.height=14, fig.width=12}
x <- params$gvizPlot

# for(i in 2:14){
#   x[[i]]@dp@pars$fontsize = 10
#   x[[i]]@name = LETTERS[i-1]
# }

plotTracks(HighlightTrack(x[1:(length(x)-1)],
                            start = params$transcript_start,
                            end = params$transcript_end,
                            chr = params$chr), col.title = 'black', 
           from = params$xrange[1], to = params$xrange[2])




```

```{r echo=FALSE, fig.height=1, fig.width=12}

if(!is.null(params$tf_legend)){
 params$tf_legend
}

```


```{r echo=FALSE, message=FALSE, warning=FALSE, fig.height=12, fig.width=12}
x <- params$rnaPlot
eval(x)
```

