---
title: "ATAC-seq Report"
output: pdf_document
params:
  gene: NA
  ymax: NA
  xrange: NA
  gvizPlot: NA
  tf_legend: NA
  rnaPlot: NA
  chr: NA
  transcriptID: NA
  ENSEMBL_Gviz_GeneTrackRegionObject: NA
---

```{r include=FALSE}

library(Gviz)
library(ggplot2)
library(ggpubr)
library(feather)
library(GenomicRanges)

```

## Summary for `r params$gene`.
Include information such as correlation cuttoff, selected transcript, etc.

\newpage

```{r echo=FALSE, fig.height=20, fig.width=12}


x <- params$gvizPlot

transcript_start <- if (params$transcriptID == 'Any') {
  getGtfCoords(params$gene, ENSEMBL_hg38_local_fromGTF)[[2]]
} else{
  filter(transcript_locations, refseq_mrna == params$transcriptID)$transcript_start
}

transcript_end <- if (params$transcriptID == 'Any') {
  getGtfCoords(params$gene, ENSEMBL_hg38_local_fromGTF)[[3]]
} else{
  filter(transcript_locations, refseq_mrna == params$transcriptID)$transcript_end
}



plotTracks(
  HighlightTrack(
    x[1:(length(x) - 1)],
    start = transcript_start,
    end = transcript_end,
    chr = params$chr,
    col = '#EEEEEE',
    fill = '#EEEEEE'
  ),
  col.title = 'black',
  from = params$xrange[1],
  to = params$xrange[2]
)


```

```{r echo=FALSE, fig.height=1, fig.width=12}

if (!is.null(params$tf_legend)) {
  params$tf_legend
}

```


```{r echo=FALSE, message=FALSE, warning=FALSE, fig.height=12, fig.width=12}

x <- params$rnaPlot
eval(x)

```

