
<!-- saved from url=(0024)http://brain.chromat.in/ -->
<html><head><meta http-equiv="Content-Type" content="text/html; charset=windows-1252">

		<style>
                     h3 {
			     font-size: 10pt
		     }
                     h2 {
			     font-size: 12pt
		     }
		     h1 {
			     font-size: 14pt
		     }
		     .floatright {
			     float: right;
		     }
		</style>
		<title>Chromatin accessibility dynamics in a model of human forebrain development</title>
	</head>
	<body>
		<h1>Chromatin accessibility dynamics in a model of human forebrain development (<a href="http://dx.doi.org/10.1126/science.aay1645">available here</a>)</h1>
		<h3>Alexandro E. Trevino<sup>*</sup>, Nasa Sinnott-Armstrong<sup>*</sup>, Jimena Andersen<sup>*</sup>, Se-Jin Yoon, Nina Huber, Jonathan K. Pritchard, Howard Y. Chang, William J. Greenleaf<sup>+</sup>, Sergiu P. Pasca<sup>+</sup> </h3>

		<h2>Abstract</h2>
		<div class="floatright"><img src="SummaryFigureImage.PNG" height=400 /></div>
                Forebrain development is characterized by highly synchronized cellular processes, the perturbing of which can cause disease. To chart the regulatory activity underlying these events, we generated amap of accessible chromatin in human forebrain organoids. To capture corticogenesis, we sampled glial and neuronal lineages from dorsal or ventral forebrain organoids over 20 months. Active chromatin regions identified in human brain tissue were observed in organoids at different developmental stages. We used this resource to map genetic risk for disease and to explore evolutionary conservation. Moreover, we integrated chromatin accessibility with transcriptomics to identify putative enhancer-gene linkages and transcription factors that regulate human corticogenesis. Overall, this platform brings insights into gene-regulatory dynamics at previously inaccessible stages of human forebrain development, including signatures of neuropsychiatric disorders.

		<h2>Data description</h2> 
		The data presented encompass the called peaks, ATAC-seq signal tracks, and RNA-seq expression from the compendium of neural spheroid samples generated as part of this study. By selecting a target gene, the accessible regions that are surrounding that gene, including those correlated with gene expression, can be visualized. Similarly, searching for a SNP enables visualization of the signal tracks and peaks around a given variant.

		<h2>Interactive session</h2>
		For users who would like to compare these data to their own studies, an interactive <a href="http://epigenomegateway.wustl.edu/browser/?sessionFile=http://brainchromatin.stanford.edu/session.json">WashU Epigenome Browser</a> session is also available.

		<h2>Data access</h2>
		Data are available on GEO under accession <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132403">GSE132403</a>. Reasonable requests for additional data and reagents should be directed to the corresponding authors.

		<h2>About</h2>

<p>This software was developed by <a target="_blank" href="https://github.com/kkovary/" >Kyle Kovary</a> and is available freely under the MIT license. The source code is available <a target="_blank" href="https://github.com/kkovary/ATAC-seq-Shiny-App">on GitHub</a>.<br/>
For any comments or questions about this web app, please <a target="_blank" href="mailto:kylekovary@protonmail.com">contact Kyle</a>.</p>

</body></html>

